### PR TITLE
Ben/add mtl

### PR DESCRIPTION
**Pull request by benleetownsend**
Needs Docs and tests, Here for reference

### CODE REVIEW

File: finetune/base.py [hunk 1]
```diff
@@ -156,14 +157,18 @@ def finetune(self, Xs, Y=None, batch_size=None):
         batch_size = batch_size or self.config.batch_size
 
         val_input_fn, train_input_fn, val_size, val_interval = self.input_pipeline.get_train_input_fns(Xs, Y, batch_size=batch_size)
-        if val_size <= 10 and self.config.keep_best_model:
-            tf.logging.warning(
-                "Early stopping / keeping best model with a validation size of {} is likely to case undesired results".format(val_size))
+        if self.config.keep_best_model:
+            if isinstance(val_size, dict):
+                tf.logging.warning("Cannot early stop or keep best model with MTL")
```
Comment (madisonmay) [inline finetune/base.py]: We shouldn't change this in this PR but we should consider thinking about how early stopping should work for an MTL problem.  I.e. mean loss across tasks or something stops decreasing.

File: finetune/comparison.py [hunk 2]
```diff
@@ -29,8 +29,8 @@ def _text_to_ids(self, pair, Y=None, pad_token=None):
 
     def feed_shape_type_def(self):
         TS = tf.TensorShape
-        return ({"tokens": tf.int32, "mask": tf.int32}, tf.int32), (
-            {"tokens": TS([2, self.config.max_length, 2]), "mask": TS([2, self.config.max_length])},
+        return ({"tokens": tf.int32, "mask": tf.float32}, tf.float32), (
+            {"tokens": TS([None, self.config.max_length, 2]), "mask": TS([None, self.config.max_length])},
```
Comment (madisonmay) [inline finetune/comparison.py]: Do we support comparison w/ > 2 items?  Curious why this change was necessary?

Comment (benleetownsend) [reply finetune/comparison.py]: I don't believe this is actually necessary. I changed it to try to trick the shape inference of datasets.

File: finetune/mtl.py [hunk 3]
```diff
@@ -58,6 +58,7 @@ def get_train_input_fns(self, Xs, Y=None, batch_size=None, val_size=None):
             input_func_normalised, val_func_normalised = get_input_fns(task_id, input_func, val_func)
             input_funcs.append(input_func_normalised)
             val_funcs[task_name] = val_func_normalised
+            val_funcs[task_name + "_train"] = lambda: input_func_normalised().take(val_sizes[task_name])
```
Comment (benleetownsend) [inline finetune/mtl.py]: This is somehow broken and actually evals for the combined dataset. Not sure how or why. Needs attention

Comment (benleetownsend) [reply finetune/mtl.py]: Scratch that, doesn't eval for the combined dataset, evals just for the final task, due to input_func_normalised changing. Must define as closure.

Comment (benleetownsend) [reply finetune/mtl.py]: Fixed

File: finetune/model.py [hunk 4]
```diff
@@ -131,15 +140,19 @@ def _model_fn(features, labels, mode, params):
 
         if mode == tf.estimator.ModeKeys.TRAIN:
             total_num_steps = params.n_epochs * params.dataset_size // params.batch_size
-            lr_decay = lambda lr, global_step: lr * schedules[params.lr_schedule](
-                tf.to_float(global_step) / total_num_steps
+            lr_decay = lambda lr, global_step: tf.maximum(0.0,
```
Comment (madisonmay) [inline finetune/model.py]: It kind of worries me that we need this but I don't think it can hurt.

File: finetune/mtl.py [hunk 5]
```diff
@@ -0,0 +1,152 @@
+import tensorflow as tf
+from finetune.base import BaseModel
+from finetune.input_pipeline import BasePipeline
+from finetune.errors import FinetuneError
+
+from finetune.sequence_labeling import SequenceLabeler
+from finetune.utils import indico_to_finetune_sequence
+
+
+def get_input_fns(tid, i_fn, v_fn):
```
Comment (madisonmay) [inline finetune/mtl.py]: Would appreciate more descriptive variable names.  I get that these are meant to be input_fn and validation_fn but still think it's worth making things more explicit.

File: finetune/mtl.py [hunk 6]
```diff
@@ -0,0 +1,152 @@
...
+        print(self.config.task_name_to_id)
+
+    def _get_input_pipeline(self):
+        return MultiTaskPipeline(self.config)
+
+    def featurize(self, X):
+        return super().featurize(X)
+
+    def predict(self, X):
+        raise FinetuneError("Predict is not implemented yet for MTL")
```
Comment (madisonmay) [inline finetune/mtl.py]: Feel like we could sensically support this using the same format as fit / finetune.

File: finetune/mtl.py [hunk 7]
```diff
@@ -0,0 +1,152 @@
...
+
+            return loss_tensor
+    return tf.equal(task_id_i, config.task_name_to_id[task]), loss
+
+
+class MultiTask(BaseModel):
+
+    def __init__(self, tasks, **kwargs):
+        super().__init__(**kwargs)
+        self._is_seq_task = [task_name for task_name, t in tasks.items() if t == SequenceLabeler]
```
Comment (madisonmay) [inline finetune/mtl.py]: Don't feel like this needs to be an attribute -- could just check this in the finetune loop.

File: finetune/mtl.py [hunk 8]
```diff
@@ -0,0 +1,173 @@
...
+                continue
+            pred_model = ModelClass()
+            pred_model.input_pipeline = self.config.task_input_pipelines[name]
+            pred_model.saver.variables = {
+                k.replace("/target_model_{}".format(name), ""): v for k, v in self.saver.variables.items()
+            }
+            predictions[name] = pred_model.predict(X[name])
+        return predictions
+
+    def predict_proba(self, X):
```
Comment (madisonmay) [inline finetune/mtl.py]: This should follow the same format, right?  Maybe w/ a fallback if this is unsupported by the task model?

Comment (madisonmay) [reply finetune/mtl.py]: Works for me
